src/mds/StrayManager.cc

```diff
@@ -0,0 +1,863 @@
+    ops_avail = max_purge_ops - ops_in_flight;
+  } else {
+    ops_avail = 0;
+  }
+
+  /* The ops_in_flight > 0 condition here handles the case where the
+   * ops required by this inode would never fit in the limit: we wait
+   * instead until nothing else is running */
+  if (ops_in_flight > 0 && ops_avail < ops_required) {
+    dout(20) << __func__ << ": throttling on max ops (require "
+             << ops_required << ", " << ops_in_flight << " in flight" << dendl;
+    return false;
+  }
+
+  // Resources are available, acquire them and execute the purge
+  files_purging += 1;
+  dout(10) << __func__ << ": allocating allowance "
+    << ops_required << " to " << ops_in_flight << " in flight" << dendl;
+  ops_in_flight += ops_required;
+  logger->set(l_mdc_num_purge_ops, ops_in_flight);
+  if (trunc) {
+    truncate(dn, ops_required);
+  } else {
+    purge(dn, ops_required);
+  }
+  return true;
+}
+
+uint32_t StrayManager::_calculate_ops_required(CInode *in, bool trunc)
+{
+  uint32_t ops_required = 0;
+  if (in->is_dir()) {
+    // Directory, count dirfrags to be deleted
+    std::list<frag_t> ls;
+    if (!in->dirfragtree.is_leaf(frag_t())) {
+      in->dirfragtree.get_leaves(ls);
+    }
+    // One for the root, plus any leaves
+    ops_required = 1 + ls.size();
+  } else {
+    // File, work out concurrent Filer::purge deletes
+    const uint64_t period = (uint64_t)in->inode.layout.fl_object_size *
+		      (uint64_t)in->inode.layout.fl_stripe_count;
+    const uint64_t to = MAX(in->inode.max_size_ever,
+            MAX(in->inode.size, in->inode.get_max_size()));
+
+    const uint64_t num = MAX(1, (to + period - 1) / period);
+    ops_required = MIN(num, g_conf->filer_max_purge_ops);
+
+    // Account for removing (or zeroing) backtrace
+    ops_required += 1;
+
+    // Account for deletions for old pools
+    if (!trunc) {
+      ops_required += in->get_projected_inode()->old_pools.size();
+    }
+  }
+
+  return ops_required;
+}
+
+void StrayManager::advance_delayed()
+{
+  for (elist<CDentry*>::iterator p = delayed_eval_stray.begin(); !p.end(); ) {
+    CDentry *dn = *p;
+    ++p;
+    dn->item_stray.remove_myself();
+    num_strays_delayed--;
+
+    if (dn->get_projected_linkage()->is_null()) {
+      /* A special case: a stray dentry can go null if its inode is being
+       * re-linked into another MDS's stray dir during a shutdown migration. */
+      dout(4) << __func__ << ": delayed dentry is now null: " << *dn << dendl;
+      continue;
+    }
+
+    const bool purging = eval_stray(dn);
+    if (!purging) {
+      derr << "Dentry " << *dn << " was purgeable but no longer is!" << dendl;
+      /*
+       * This can happen if a stray is purgeable, but has gained an extra
+       * reference by virtue of having its backtrace updated.
+       * FIXME perhaps we could simplify this further by
+       * avoiding writing the backtrace of purge-ready strays, so
+       * that this code could be more rigid?
```

Comment:
> I'm just unclear on how we could have a dirty parent when we're purgeable anyway. Does that just happen if the inode gets renamed and then deleted quickly enough?

I don't remember -- this was a case that I saw in testing, which tripped an `assert(purging)` that I had originally put in here, so I changed the assert to log instead (and added the comment).

> And is having the checking machinery to block it here really simpler than simply clearing the dirty parent field as soon as it gets marked stray? (I guess the downside there would be that fsck might have a harder time with such things, but I don't think that's a big deal...)

I wasn't claiming that it was simpler to have these checks here -- I'd like to remove them too (this particular piece of logic is inherited from where it used to live in MDCache::trim, where it would simply invoke eval_stray for everything in the delayed list and not pay attention to the outcome).

> I don't really understand your comment about testing separately.

This particular bit of logic is something that deals with a case that doesn't reproduce all that reliably (did not happen every fs suite run), so if we change it I'd like to do that in a separate PR after merging this one, and give that other PR several test runs.